File: python/tests/scenarios.py
from __future__ import annotations

import asyncio
import base64
import json
import time
from typing import Literal, TypedDict

import websockets

from .helpers import make_client, wait_for
from .mock_server import (
    FIXED_ATTACHMENT_ID,
    FIXED_SESSION_ID,
    MockServer,
    start_mock_server,
)

ScenarioName = Literal[
    "normal_chat",
    "streaming_chat",
    "reconnect_resync",
    "auth_refresh",
    "file_upload",
    "liveness",
]


class ScenarioMessage(TypedDict, total=False):
    type: str
    seq: int


class ScenarioResult(TypedDict):
    scenario: str
    messages_received: list[ScenarioMessage]
    session_state_final: str
    channel_state_final: str
    errors_received: list[str]
    callback_call_count: int


class ScenarioExecution(TypedDict):
    result: ScenarioResult
    schema_violations: list[dict[str, object]]
    diagnostics: dict[str, int]


def _make_jwt(claims: dict[str, object]) -> str:
    header = base64.urlsafe_b64encode(
        json.dumps({"alg": "HS256", "typ": "JWT"}).encode()
    ).rstrip(b"=").decode()
    payload = base64.urlsafe_b64encode(json.dumps(claims).encode()).rstrip(b"=").decode()
    return f"{header}.{payload}.fakesig"


def _snapshot_result(
    scenario: ScenarioName,
    client,
    received: list[dict[str, object]],
) -> ScenarioResult:
    return {
        "scenario": scenario,
        "messages_received": [
            {
                "type": str(message["type"]),
                **({"seq": int(message["seq"])} if isinstance(message.get("seq"), int) else {}),
            }
            for message in received
        ],
        "session_state_final": client.session_state,
        "channel_state_final": client.channel_state,
        "errors_received": [
            str(message["payload"]["code"])
            for message in received
            if message["type"] == "system::error" and isinstance(message["payload"], dict)
        ],
        "callback_call_count": len(received),
    }


def _finalize_execution(
    scenario: ScenarioName,
    client,
    server: MockServer,
    received: list[dict[str, object]],
) -> ScenarioExecution:
    return {
        "result": _snapshot_result(scenario, client, received),
        "schema_violations": list(server.schema_violations),
        "diagnostics": {
            "ws_connection_count": server.ws_connection_count,
            "refresh_call_count": server.refresh_call_count,
            "upload_call_count": server.upload_call_count,
            "received_message_count": server.received_message_count,
        },
    }


async def run_normal_chat_scenario(schema_validation: bool = False) -> ScenarioExecution:
    server = await start_mock_server(
        auto_chat_answer=True,
        auto_init_echo=True,
        enable_schema_validation=schema_validation,
    )
    received: list[dict[str, object]] = []
    client = make_client(server, received)

    try:
        await client.connect()
        await wait_for(lambda: client.session_id is not None)
        await client.send_message("Hello runtime")
        await wait_for(
            lambda: any(
                msg["type"] == "chat::answer" and msg["payload"].get("answer_kind") == "final"
                for msg in received
            )
        )
        return _finalize_execution("normal_chat", client, server, received)
    finally:
        await client.disconnect()
        await server.close()


async def run_streaming_chat_scenario(schema_validation: bool = False) -> ScenarioExecution:
    stream_sent = False

    async def on_ws_message(
        msg: dict[str, object],
        ws: websockets.WebSocketServerProtocol,
        srv: MockServer,
    ) -> None:
        nonlocal stream_sent
        if msg.get("type") == "chat::message" and not stream_sent:
            stream_sent = True
            seq = [1]

            def env(type_: str, payload: dict[str, object]) -> dict[str, object]:
                envelope = {
                    "type": type_,
                    "schema": "1.0",
                    "session_id": FIXED_SESSION_ID,
                    "seq": seq[0],
                    "payload": payload,
                    "ts": "2026-04-04T00:00:00Z",
                }
                seq[0] += 1
                return envelope

            await srv.send_to(ws, env("chat::partial", {
                "content": "Quantum entanglement is ",
                "role": "assistant",
                "turn_id": "turn_002",
            }))
            await srv.send_to(ws, env("chat::partial", {
                "content": "a phenomenon where two particles ",
                "role": "assistant",
                "turn_id": "turn_002",
            }))
            await srv.send_to(ws, env("chat::answer", {
                "content": "Quantum entanglement is a phenomenon where two particles remain correlated regardless of distance.",
                "role": "assistant",
                "answer_kind": "final",
                "turn_id": "turn_002",
            }))

    server = await start_mock_server(
        auto_chat_answer=False,
        auto_init_echo=True,
        enable_schema_validation=schema_validation,
        on_ws_message=on_ws_message,
    )
    received: list[dict[str, object]] = []
    client = make_client(server, received)

    try:
        await client.connect()
        await wait_for(lambda: client.session_id is not None)
        await client.send_message("Explain quantum entanglement briefly.")
        await wait_for(
            lambda: any(
                msg["type"] == "chat::answer" and msg["payload"].get("answer_kind") == "final"
                for msg in received
            ),
            timeout=3.0,
        )
        return _finalize_execution("streaming_chat", client, server, received)
    finally:
        await client.disconnect()
        await server.close()


async def run_reconnect_resync_scenario(schema_validation: bool = False) -> ScenarioExecution:
    resynced = False

    async def on_ws_message(
        msg: dict[str, object],
        ws: websockets.WebSocketServerProtocol,
        srv: MockServer,
    ) -> None:
        nonlocal resynced
        if msg.get("type") == "system::resync":
            resynced = True
            await srv.send_to(ws, {
                "type": "chat::answer",
                "schema": "1.0",
                "session_id": FIXED_SESSION_ID,
                "seq": 1,
                "payload": {
                    "content": "Long task completed",
                    "role": "assistant",
                    "answer_kind": "final",
                    "turn_id": "turn_003",
                },
                "ts": "2026-04-04T00:00:00Z",
            })

    server = await start_mock_server(
        auto_chat_answer=False,
        auto_init_echo=True,
        enable_schema_validation=schema_validation,
        on_ws_message=on_ws_message,
    )
    received: list[dict[str, object]] = []
    client = make_client(server, received, ping_interval=60.0, stale_threshold=60.0)

    try:
        await client.connect()
        await wait_for(lambda: client.session_id is not None)
        await client.send_message("Start a long task")
        await server.drop_connections()
        await wait_for(lambda: resynced, timeout=5.0)
        await wait_for(
            lambda: any(
                msg["type"] == "chat::answer" and msg["payload"].get("answer_kind") == "final"
                for msg in received
            ),
            timeout=5.0,
        )
        return _finalize_execution("reconnect_resync", client, server, received)
    finally:
        await client.disconnect()
        await server.close()


async def run_auth_refresh_scenario(schema_validation: bool = False) -> ScenarioExecution:
    first_answer_sent = False
    second_answer_sent = False
    expiring_token = _make_jwt({"exp": int(time.time()) + 30})
    refreshed_token = _make_jwt({"exp": int(time.time()) + 3600})

    async def on_ws_message(
        msg: dict[str, object],
        ws: websockets.WebSocketServerProtocol,
        srv: MockServer,
    ) -> None:
        nonlocal first_answer_sent, second_answer_sent
        if msg.get("type") == "chat::message" and not first_answer_sent:
            first_answer_sent = True
            await srv.send_to(ws, {
                "type": "chat::answer",
                "schema": "1.0",
                "session_id": FIXED_SESSION_ID,
                "seq": 1,
                "payload": {
                    "content": "Response to first message",
                    "role": "assistant",
                    "answer_kind": "final",
                    "turn_id": "turn_004",
                },
                "ts": "2026-04-04T00:00:00Z",
            })

            async def close_later() -> None:
                await asyncio.sleep(0.025)
                await ws.close(1011, "refresh")

            asyncio.create_task(close_later())
            return

        if msg.get("type") == "chat::message" and first_answer_sent and not second_answer_sent:
            second_answer_sent = True
            await srv.send_to(ws, {
                "type": "chat::answer",
                "schema": "1.0",
                "session_id": FIXED_SESSION_ID,
                "seq": 2,
                "payload": {
                    "content": "Response to second message",
                    "role": "assistant",
                    "answer_kind": "final",
                    "turn_id": "turn_005",
                },
                "ts": "2026-04-04T00:00:00Z",
            })

    server = await start_mock_server(
        auto_chat_answer=False,
        auto_init_echo=True,
        enable_schema_validation=schema_validation,
        initial_access_token=expiring_token,
        refreshed_access_token=refreshed_token,
        on_ws_message=on_ws_message,
    )
    received: list[dict[str, object]] = []
    client = make_client(
        server,
        received,
        ping_interval=60.0,
        stale_threshold=60.0,
        resync_timeout=2.0,
    )

    try:
        await client.connect()
        await wait_for(lambda: client.session_id is not None)
        await client.send_message("First message before refresh")
        await wait_for(
            lambda: any(
                msg["type"] == "chat::answer" and msg["payload"].get("turn_id") == "turn_004"
                for msg in received
            ),
            timeout=3.0,
        )
        await wait_for(lambda: server.refresh_call_count >= 1, timeout=5.0)
        await wait_for(lambda: server.ws_connection_count >= 2, timeout=5.0)
        await client.send_message("Second message after refresh")
        await wait_for(
            lambda: any(
                msg["type"] == "chat::answer" and msg["payload"].get("turn_id") == "turn_005"
                for msg in received
            ),
            timeout=3.0,
        )
        return _finalize_execution("auth_refresh", client, server, received)
    finally:
        await client.disconnect()
        await server.close()


async def run_file_upload_scenario(schema_validation: bool = False) -> ScenarioExecution:
    async def on_ws_message(
        msg: dict[str, object],
        ws: websockets.WebSocketServerProtocol,
        srv: MockServer,
    ) -> None:
        if msg.get("type") == "chat::message":
            await srv.send_to(ws, {
                "type": "chat::answer",
                "schema": "1.0",
                "session_id": FIXED_SESSION_ID,
                "seq": 1,
                "payload": {
                    "content": "The document covers three main topics.",
                    "role": "assistant",
                    "answer_kind": "final",
                    "turn_id": "turn_006",
                },
                "ts": "2026-04-04T00:00:00Z",
            })

    server = await start_mock_server(
        auto_chat_answer=False,
        auto_init_echo=True,
        enable_schema_validation=schema_validation,
        on_ws_message=on_ws_message,
    )
    received: list[dict[str, object]] = []
    client = make_client(server, received)

    try:
        await client.connect()
        await wait_for(lambda: client.session_id is not None)
        attachment_id = await client.upload_attachment(b"\x00\x01\x02\x03")
        if attachment_id != FIXED_ATTACHMENT_ID:
            raise AssertionError(f"unexpected attachment id {attachment_id}")
        await client.send_message(
            "Please summarize this document.",
            attachments=[attachment_id],
        )
        await wait_for(
            lambda: any(
                msg["type"] == "chat::answer" and msg["payload"].get("answer_kind") == "final"
                for msg in received
            ),
            timeout=3.0,
        )
        return _finalize_execution("file_upload", client, server, received)
    finally:
        await client.disconnect()
        await server.close()


async def run_liveness_scenario(schema_validation: bool = False) -> ScenarioExecution:
    ping_count = 0

    async def on_ws_message(msg: dict[str, object], ws, srv) -> None:
        nonlocal ping_count
        if msg.get("type") == "system::ping":
            ping_count += 1

    server = await start_mock_server(
        auto_pong=True,
        auto_init_echo=True,
        enable_schema_validation=schema_validation,
        on_ws_message=on_ws_message,
    )
    received: list[dict[str, object]] = []
    client = make_client(
        server,
        received,
        ping_interval=0.1,
        pong_timeout=0.5,
        stale_threshold=5.0,
    )

    try:
        await client.connect()
        await wait_for(lambda: client.session_id is not None)
        await wait_for(lambda: ping_count >= 2, timeout=3.0)
        await asyncio.sleep(0.05)
        return _finalize_execution("liveness", client, server, received)
    finally:
        await client.disconnect()
        await server.close()


async def run_scenario_by_name(
    scenario: ScenarioName,
    *,
    schema_validation: bool = False,
) -> ScenarioExecution:
    if scenario == "normal_chat":
        return await run_normal_chat_scenario(schema_validation=schema_validation)
    if scenario == "streaming_chat":
        return await run_streaming_chat_scenario(schema_validation=schema_validation)
    if scenario == "reconnect_resync":
        return await run_reconnect_resync_scenario(schema_validation=schema_validation)
    if scenario == "auth_refresh":
        return await run_auth_refresh_scenario(schema_validation=schema_validation)
    if scenario == "file_upload":
        return await run_file_upload_scenario(schema_validation=schema_validation)
    if scenario == "liveness":
        return await run_liveness_scenario(schema_validation=schema_validation)
    raise AssertionError(f"unknown scenario {scenario}")


SCENARIO_NAMES: list[ScenarioName] = [
    "normal_chat",
    "streaming_chat",
    "reconnect_resync",
    "auth_refresh",
    "file_upload",
    "liveness",
]

File: js/tests/scenarios.ts
import { WebSocket } from 'ws';

import { startMockServer, FIXED_ATTACHMENT_ID, FIXED_SESSION_ID, type MockServer } from './mock-server.js';
import { makeClient } from './test-client-factory.js';
import { waitFor } from './helpers.js';
import type { CortexMessage } from '../src/types.js';
import type { SchemaViolation } from './schema-validation.js';

export type ScenarioName =
  | 'normal_chat'
  | 'streaming_chat'
  | 'reconnect_resync'
  | 'auth_refresh'
  | 'file_upload'
  | 'liveness';

export interface ScenarioResult {
  scenario: string;
  messages_received: Array<{ type: string; seq?: number }>;
  session_state_final: string;
  channel_state_final: string;
  errors_received: string[];
  callback_call_count: number;
}

export interface ScenarioExecution {
  result: ScenarioResult;
  schema_violations: SchemaViolation[];
  diagnostics: {
    ws_connection_count: number;
    refresh_call_count: number;
    upload_call_count: number;
    received_message_count: number;
  };
}

export interface ScenarioOptions {
  schemaValidation?: boolean;
}

function snapshotResult(
  scenario: ScenarioName,
  client: ReturnType<typeof makeClient>,
  received: CortexMessage[],
): ScenarioResult {
  return {
    scenario,
    messages_received: received.map((message) => ({
      type: message.type,
      seq: typeof message.seq === 'number' ? message.seq : undefined,
    })),
    session_state_final: client.sessionState,
    channel_state_final: client.channelState,
    errors_received: received
      .filter((message) => message.type === 'system::error')
      .map((message) => String(message.payload['code'])),
    callback_call_count: received.length,
  };
}

function finalizeExecution(
  scenario: ScenarioName,
  client: ReturnType<typeof makeClient>,
  server: MockServer,
  received: CortexMessage[],
): ScenarioExecution {
  return {
    result: snapshotResult(scenario, client, received),
    schema_violations: [...server.schemaViolations],
    diagnostics: {
      ws_connection_count: server.wsConnectionCount,
      refresh_call_count: server.refreshCallCount,
      upload_call_count: server.uploadCallCount,
      received_message_count: server.receivedMessageCount,
    },
  };
}

function makeJwt(claims: Record<string, unknown>): string {
  const header = Buffer.from(JSON.stringify({ alg: 'HS256', typ: 'JWT' })).toString('base64url');
  const payload = Buffer.from(JSON.stringify(claims)).toString('base64url');
  return `${header}.${payload}.fakesig`;
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function runNormalChatScenario(options: ScenarioOptions = {}): Promise<ScenarioExecution> {
  const server = await startMockServer({
    autoChatAnswer: true,
    autoInitEcho: true,
    enableSchemaValidation: options.schemaValidation,
  });
  const received: CortexMessage[] = [];
  const client = makeClient(server, (message) => received.push(message));

  try {
    await client.connect();
    await waitFor(() => client.sessionId !== null);
    await client.sendMessage({ content: 'Hello runtime' });
    await waitFor(() => received.some((message) => message.type === 'chat::answer' && message.payload['answer_kind'] === 'final'));
    return finalizeExecution('normal_chat', client, server, received);
  } finally {
    await client.disconnect();
    await server.close();
  }
}

export async function runStreamingChatScenario(options: ScenarioOptions = {}): Promise<ScenarioExecution> {
  let streamSent = false;
  const server = await startMockServer({
    autoChatAnswer: false,
    autoInitEcho: true,
    enableSchemaValidation: options.schemaValidation,
    onWsMessage: (parsed, ws, srv) => {
      if (parsed['type'] === 'chat::message' && !streamSent) {
        streamSent = true;
        let seq = 1;
        const makeEnv = (type: string, payload: Record<string, unknown>) => ({
          type,
          schema: '1.0',
          session_id: FIXED_SESSION_ID,
          seq: seq++,
          payload,
          ts: new Date().toISOString(),
        });
        srv.sendTo(ws, makeEnv('chat::partial', {
          content: 'Quantum entanglement is ',
          role: 'assistant',
          turn_id: 'turn_002',
        }));
        srv.sendTo(ws, makeEnv('chat::partial', {
          content: 'a phenomenon where two particles ',
          role: 'assistant',
          turn_id: 'turn_002',
        }));
        srv.sendTo(ws, makeEnv('chat::answer', {
          content: 'Quantum entanglement is a phenomenon where two particles remain correlated regardless of distance.',
          role: 'assistant',
          answer_kind: 'final',
          turn_id: 'turn_002',
        }));
      }
    },
  });
  const received: CortexMessage[] = [];
  const client = makeClient(server, (message) => received.push(message));

  try {
    await client.connect();
    await waitFor(() => client.sessionId !== null);
    await client.sendMessage({ content: 'Explain quantum entanglement briefly.' });
    await waitFor(() => received.some((message) => message.type === 'chat::answer' && message.payload['answer_kind'] === 'final'), 3000);
    return finalizeExecution('streaming_chat', client, server, received);
  } finally {
    await client.disconnect();
    await server.close();
  }
}

export async function runReconnectResyncScenario(options: ScenarioOptions = {}): Promise<ScenarioExecution> {
  let resynced = false;
  const server = await startMockServer({
    autoInitEcho: true,
    autoChatAnswer: false,
    enableSchemaValidation: options.schemaValidation,
    onWsMessage: (parsed, ws, srv) => {
      if (parsed['type'] === 'system::resync') {
        resynced = true;
        srv.sendTo(ws, {
          type: 'chat::answer',
          schema: '1.0',
          session_id: FIXED_SESSION_ID,
          seq: 1,
          payload: {
            content: 'Long task completed',
            role: 'assistant',
            answer_kind: 'final',
            turn_id: 'turn_003',
          },
          ts: new Date().toISOString(),
        });
      }
    },
  });
  const received: CortexMessage[] = [];
  const client = makeClient(server, (message) => received.push(message), {
    pingInterval: 60000,
    staleThreshold: 60000,
  });

  try {
    await client.connect();
    await waitFor(() => client.sessionId !== null);
    await client.sendMessage({ content: 'Start a long task' });
    server.dropConnections();
    await waitFor(() => resynced, 5000);
    await waitFor(() => received.some((message) => message.type === 'chat::answer' && message.payload['answer_kind'] === 'final'), 5000);
    return finalizeExecution('reconnect_resync', client, server, received);
  } finally {
    await client.disconnect();
    await server.close();
  }
}

export async function runAuthRefreshScenario(options: ScenarioOptions = {}): Promise<ScenarioExecution> {
  let firstAnswerSent = false;
  let secondAnswerSent = false;
  const expiringToken = makeJwt({ exp: Math.floor(Date.now() / 1000) + 30 });
  const refreshedToken = makeJwt({ exp: Math.floor(Date.now() / 1000) + 3600 });

  const server = await startMockServer({
    autoInitEcho: true,
    autoChatAnswer: false,
    enableSchemaValidation: options.schemaValidation,
    initialAccessToken: expiringToken,
    refreshedAccessToken: refreshedToken,
    onWsMessage: (parsed, ws, srv) => {
      if (parsed['type'] === 'chat::message' && !firstAnswerSent) {
        firstAnswerSent = true;
        srv.sendTo(ws, {
          type: 'chat::answer',
          schema: '1.0',
          session_id: FIXED_SESSION_ID,
          seq: 1,
          payload: {
            content: 'Response to first message',
            role: 'assistant',
            answer_kind: 'final',
            turn_id: 'turn_004',
          },
          ts: new Date().toISOString(),
        });
        setTimeout(() => {
          if (ws.readyState === WebSocket.OPEN) {
            ws.close(1011, 'refresh');
          }
        }, 25);
        return;
      }

      if (parsed['type'] === 'chat::message' && firstAnswerSent && !secondAnswerSent) {
        secondAnswerSent = true;
        srv.sendTo(ws, {
          type: 'chat::answer',
          schema: '1.0',
          session_id: FIXED_SESSION_ID,
          seq: 2,
          payload: {
            content: 'Response to second message',
            role: 'assistant',
            answer_kind: 'final',
            turn_id: 'turn_005',
          },
          ts: new Date().toISOString(),
        });
      }
    },
  });
  const received: CortexMessage[] = [];
  const client = makeClient(server, (message) => received.push(message), {
    pingInterval: 60000,
    staleThreshold: 60000,
    resyncTimeout: 2000,
  });

  try {
    await client.connect();
    await waitFor(() => client.sessionId !== null);
    await client.sendMessage({ content: 'First message before refresh' });
    await waitFor(() => received.some((message) => message.type === 'chat::answer' && message.payload['turn_id'] === 'turn_004'), 3000);
    await waitFor(() => server.refreshCallCount >= 1, 5000);
    await waitFor(() => server.wsConnectionCount >= 2, 5000);
    await client.sendMessage({ content: 'Second message after refresh' });
    await waitFor(() => received.some((message) => message.type === 'chat::answer' && message.payload['turn_id'] === 'turn_005'), 3000);
    return finalizeExecution('auth_refresh', client, server, received);
  } finally {
    await client.disconnect();
    await server.close();
  }
}

export async function runFileUploadScenario(options: ScenarioOptions = {}): Promise<ScenarioExecution> {
  const server = await startMockServer({
    autoInitEcho: true,
    autoChatAnswer: false,
    enableSchemaValidation: options.schemaValidation,
    onWsMessage: (parsed, ws, srv) => {
      if (parsed['type'] === 'chat::message') {
        srv.sendTo(ws, {
          type: 'chat::answer',
          schema: '1.0',
          session_id: FIXED_SESSION_ID,
          seq: 1,
          payload: {
            content: 'The document covers three main topics.',
            role: 'assistant',
            answer_kind: 'final',
            turn_id: 'turn_006',
          },
          ts: new Date().toISOString(),
        });
      }
    },
  });
  const received: CortexMessage[] = [];
  const client = makeClient(server, (message) => received.push(message));

  try {
    await client.connect();
    await waitFor(() => client.sessionId !== null);
    const attachmentId = await client.uploadAttachment(new Uint8Array([1, 2, 3, 4]));
    if (attachmentId !== FIXED_ATTACHMENT_ID) {
      throw new Error(`unexpected attachment id ${attachmentId}`);
    }
    await client.sendMessage({
      content: 'Please summarize this document.',
      attachments: [attachmentId],
    });
    await waitFor(() => received.some((message) => message.type === 'chat::answer' && message.payload['answer_kind'] === 'final'), 3000);
    return finalizeExecution('file_upload', client, server, received);
  } finally {
    await client.disconnect();
    await server.close();
  }
}

export async function runLivenessScenario(options: ScenarioOptions = {}): Promise<ScenarioExecution> {
  let pingSentCount = 0;
  const server = await startMockServer({
    autoInitEcho: true,
    autoPong: true,
    autoChatAnswer: false,
    enableSchemaValidation: options.schemaValidation,
    onWsMessage: (parsed) => {
      if (parsed['type'] === 'system::ping') {
        pingSentCount++;
      }
    },
  });
  const received: CortexMessage[] = [];
  const client = makeClient(server, (message) => received.push(message), {
    pingInterval: 100,
    pongTimeout: 500,
    staleThreshold: 5000,
  });

  try {
    await client.connect();
    await waitFor(() => client.sessionId !== null);
    await waitFor(() => pingSentCount >= 2, 3000);
    await delay(50);
    return finalizeExecution('liveness', client, server, received);
  } finally {
    await client.disconnect();
    await server.close();
  }
}

export async function runScenarioByName(
  scenario: ScenarioName,
  options: ScenarioOptions = {},
): Promise<ScenarioExecution> {
  switch (scenario) {
    case 'normal_chat':
      return runNormalChatScenario(options);
    case 'streaming_chat':
      return runStreamingChatScenario(options);
    case 'reconnect_resync':
      return runReconnectResyncScenario(options);
    case 'auth_refresh':
      return runAuthRefreshScenario(options);
    case 'file_upload':
      return runFileUploadScenario(options);
    case 'liveness':
      return runLivenessScenario(options);
    default:
      throw new Error(`Unknown scenario ${(scenario as never)}`);
  }
}

export const SCENARIO_NAMES: ScenarioName[] = [
  'normal_chat',
  'streaming_chat',
  'reconnect_resync',
  'auth_refresh',
  'file_upload',
  'liveness',
];
